src/entry.rs

```diff
@@ -2,211 +2,376 @@ use alloc::boxed::Box;
 use alloc::sync::Arc;
 use core::marker::PhantomData;
 use core::mem::ManuallyDrop;
+use core::ops::{Deref, Not};
 
-use crate::lock::XLock;
-use crate::node::{ReadWrite, XNode};
-use crate::xarray::SLOT_SIZE;
+use crate::node::{TryClone, XNode};
 
-/// A trait that should be implemented for the types users wish to store in an `XArray`.
-/// Items stored in an XArray are required to be 8 bytes in size, Currently it can be various pointer types.
+/// A trait for the types users wish to store in an `XArray`.
+///
+/// Items stored in an `XArray` must be representable by a `*const ()` aligned to 4. We prefer
+/// `*const ()` than `usize` to make the implementation conform to [Strict Provenance][1].
+///
+///  [1]: https://doc.rust-lang.org/std/ptr/index.html#strict-provenance
 ///
 /// # Safety
-/// Users must ensure that the produced `usize` of `into_raw()` meets the requirements for an item entry
-/// in the XArray. Specifically, if the original type is a pointer, the last two bits should be 00;
-/// if the original type is a value like usize, the last bit should be 1 (TODO).
+///
+/// Users must ensure that [`ItemEntry::into_raw()`] always produce `*const ()`s that meet the
+/// above alignment requirements.
+///
+/// Users must also ensure that as long as the value does not get dropped (e.g., by dropping the
+/// value obtaining from [`ItemEntry::from_raw()`]), it is safe to invoke [`ItemEntry::raw_as_ref`]
+/// multiple times to obtain values of [`ItemEntry::Ref`] that behave just like shared references
+/// to the underleying data.
 pub unsafe trait ItemEntry {
-    /// Converts the original type into a `usize`, consuming the ownership of the original type.
+    /// The type that behaves just like a shared references to the underleying data.
+    type Ref<'a>: Deref<Target = Self>
+    where
+        Self: 'a;
+
+    /// Converts the original value into a `*const ()`, consuming the ownership of the original
+    /// value.
+    fn into_raw(self) -> *const ();
+
+    /// Recovers the original value from a `*const ()`, reclaiming the ownership of the original
+    /// value.
     ///
-    /// This `usize` should be directly stored in an XArray's XEntry.
-    fn into_raw(self) -> usize;
+    /// # Safety
+    ///
+    /// The original value must have not been dropped, and all references obtained from
+    /// [`ItemEntry::raw_as_ref`] must be dead.
+    unsafe fn from_raw(raw: *const ()) -> Self;
 
-    /// Recovers the original type from a usize, reclaiming ownership.
+    /// Obtains a shared reference to the original value.
     ///
     /// # Safety
-    /// The raw value passed must have been produced by the corresponding `into_raw` method in this trait
-    /// from the same type.
-    unsafe fn from_raw(raw: usize) -> Self;
+    ///
+    /// The original value must outlive the lifetime parameter `'a`, and during `'a` no mutable
+    /// references to the value will exist.
+    unsafe fn raw_as_ref<'a>(raw: *const ()) -> Self::Ref<'a>;
+}
```

Comment:
After reading the discussion on Zulip, I can understand why is that necessary here, but I have a concern that this introduces a burden to users since they need additional efforts to declare a `Ref` struct when using a new type (e.g. `VmFrame`) as an `ItemEntry`. But I can't think of a better way either.

The primary issue with using `transmute` on `Arc` right now is that `Arc` is not marked for transmutation, so there's uncertainty about whether future compiler optimizations might lead to a memory layout that is inconsistent with the inner pointer. I personally think that a `transmute` annotation would be quite reasonable for `Arc`, and perhaps this could be addressed in a future version of Rust.